sdktests/server_side_eval_bucketing.go

```diff
@@ -0,0 +1,482 @@
+
+		if rolloutKind != ldmodel.RolloutKindExperiment {
+			segmentResult := evaluateFlagDetail(t, client, flagForSegmentKey, makeContext(p, false), defaultValue)
+			m.In(t).For(p.describe()+" (in segment)").Assert(segmentResult, expectedSegmentResult)
+		}
+	}
+
+	doTests := func(
+		t *ldtest.T,
+		params []bucketingTestParams,
+		rolloutKind ldmodel.RolloutKind,
+		contextKind ldcontext.Kind,
+		bucketBy ldattr.Ref,
+		makeContext func(p bucketingTestParams, shouldMatchRule bool) ldcontext.Context,
+	) {
+		for _, p := range params {
+			doTest(t, p, rolloutKind, contextKind, bucketBy, makeContext)
+		}
+	}
+
+	t.Run("basic bucketing calculations", func(t *ldtest.T) {
+		makeContext := func(p bucketingTestParams, shouldMatchRule bool) ldcontext.Context {
+			b := ldcontext.NewBuilder(ldvalue.CopyArbitraryValue(p.contextValue).StringValue())
+			if shouldMatchRule {
+				b.SetBool(matchRuleAttr, true)
+			}
+			return b.Build()
+		}
+
+		t.Run("rollouts", func(t *ldtest.T) {
+			doTests(t, makeBucketingTestParams(), ldmodel.RolloutKindRollout, ldcontext.DefaultKind, ldattr.Ref{}, makeContext)
+		})
+
+		t.Run("experiments", func(t *ldtest.T) {
+			doTests(t, makeBucketingTestParamsForExperiments(), ldmodel.RolloutKindExperiment,
+				ldcontext.DefaultKind, ldattr.Ref{}, makeContext)
+		})
+	})
+
+	t.Run("selection of context", func(t *ldtest.T) {
+		for _, multi := range []bool{false, true} {
+			desc := selectString(multi, "multi-kind", "single-kind")
+			contextKind := ldcontext.Kind("org")
+
+			t.Run(desc, func(t *ldtest.T) {
+				makeContext := func(p bucketingTestParams, shouldMatchRule bool) ldcontext.Context {
+					b := ldcontext.NewBuilder(p.contextValue).Kind(contextKind)
+					if shouldMatchRule {
+						b.SetBool(matchRuleAttr, true)
+					}
+					if multi {
+						return ldcontext.NewMulti(ldcontext.NewWithKind("wrongkind", "wrongkey"), b.Build())
+					} else {
+						return b.Build()
+					}
+				}
+
+				t.Run(desc, func(t *ldtest.T) {
+					t.Run("rollouts", func(t *ldtest.T) {
+						doTests(t, makeBucketingTestParams(), ldmodel.RolloutKindRollout, contextKind, ldattr.Ref{}, makeContext)
+					})
+
+					t.Run("experiments", func(t *ldtest.T) {
+						doTests(t, makeBucketingTestParamsForExperiments(), ldmodel.RolloutKindExperiment,
+							contextKind, ldattr.Ref{}, makeContext)
+					})
+				})
+			})
+		}
+	})
+
+	t.Run("secondary key", func(t *ldtest.T) {
+		for _, secondary := range []string{"abcdef", ""} {
+			makeContext := func(p bucketingTestParams, shouldMatchRule bool) ldcontext.Context {
+				b := ldcontext.NewBuilder(p.contextValue)
+				b.Secondary(secondary)
+				if shouldMatchRule {
+					b.SetBool(matchRuleAttr, true)
+				}
+				return b.Build()
+			}
+
+			desc := selectString(secondary == "", "secondary key is an empty string", "secondary key is a non-empty string")
+			t.Run(desc, func(t *ldtest.T) {
+				t.Run("affects bucketing calculations in rollouts", func(t *ldtest.T) {
```

Comment:
Here I removed the `NonCritical` designation in the case where secondary is `""`. U2C-aware SDKs are required to be more consistent in this regard.